### PR TITLE
Html report

### PR DESCRIPTION
**Pull request by willrogers**
@nickbattam do your worst

### CODE REVIEW

File: html_versions.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+import subprocess
+import ConfigParser
+import logging as log
+LOG_FORMAT = '%(levelname)s:  %(message)s'
+LOG_LEVEL = log.WARNING
+log.basicConfig(format=LOG_FORMAT, level=LOG_LEVEL)
+
+SUPPORT = '/dls_sw/prod/R3.14.12.3/support'
+IOC = '/dls_sw/prod/R3.14.12.3/ioc'
+RESOURCES_DIR = 'res'
```
Comment (nickbattam) [inline html_versions.py:26]: You use "/dls_sw/prod/R3.14.12.3" again later. Might be cleaner to define BASE_PATH and the SUPPORT and IOC areas as strings

File: html_versions.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        self.requested = requested
+        self.latest_release = latest_release
+        self.config_version = config_version
+        self.launcher_version = None
+        self.cfg_ioc_version = None
+        self.deps = {}
+        self.version_class = ModDetails.OK
+        self.lversion_class = ModDetails.OK
+        self.cversion_class = ModDetails.OK
+        self.rversion_class = ModDetails.OK
```
Comment (nickbattam) [inline html_versions.py]: version_class, lversion_class, rversion_class and c_version_class are uninformative variable names

File: html_versions.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    except (ConfigParser.NoSectionError, ConfigParser.NoOptionError):
+        value = None
+    return value
+
+
+def get_versions(module_cfg, coords):
+    try:
+        latest_release = utils.get_latest_version(coordinates.as_path(coords, False))
+    except ValueError:
+        latest_release = '99999'
```
Comment (nickbattam) [inline html_versions.py]: '99999' would be better as a meaningfully named constant

File: html_versions.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    return ioc_names
+
+
+def find_support_modules():
+    """Returns a list of support module names e.g. 'diagOpi' """
+    return os.listdir(SUPPORT)
+
+
+def handle_one_module(module_cfg, module_name, launcher_version, cfg_ioc_version, area):
+    coords = coordinates.ModCoord('/dls_sw/prod/R3.14.12.3/', area, module_name, None)
```
Comment (nickbattam) [inline html_versions.py]: Reuse of BASE_PATH

File: html_versions.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+
+
+class ModDetails(object):
+    # CSS classes
+    OK = 'ok'
+    OUT_OF_DATE = 'out-of-date'
+    CONFIGURED = 'configured'
+
+    def __init__(self, name, requested, latest_release, config_version):
+        self.name = name
```
Comment (nickbattam) [inline html_versions.py:42]: Constructor is clumsy; you use it twice both times setting additional state in the following lines. Would creating two constructors tailored to the two use-cases (dependency or top-level module) be simpler to use?

File: html_versions.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    env = jinja2.Environment(loader=jinja2.PackageLoader(script_name,
+                                                         RESOURCES_DIR))
+    template = env.get_template(HTML_TEMPLATE)
+    header_tags = ['thead', 'tfoot']
+    full_html = template.render(header_tags=header_tags, nheaders=max_deps,
+                                mod_details=sorted_details)
+
+    with open(REPORT, 'w') as f:
+        f.write(full_html)
+    print('Created HTML report: {}'.format(REPORT))
```
Comment (nickbattam) [inline html_versions.py:186]: What happens if an exception is thrown by any part of this open/write?

File: html_versions.py [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+    return launcher_versions
+
+
+def get_configure_ioc_versions(ioc_names):
+    """Determine where possible versions of modules used in configure-ioc.
+
+    Returns:
+        dict: module name => version string or 'work'
+    """
+    cfg_ioc = subprocess.check_output(CFG_IOC_CMD).strip().split('\n')
```
Comment (nickbattam) [inline html_versions.py:211]: Need to handle CalledProcessErrors raised by the subprocess call

File: html_versions.py [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+    ioc_paths = [path for _, path in (line.split() for line in cfg_ioc)]
```
Comment (nickbattam) [inline html_versions.py:212]: unclear what this line is doing

File: html_versions.py [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+    for ioc_name in ioc_names:
+        for ioc_path in ioc_paths:
+            if ioc_name in ioc_path:
+                if '/work/' in ioc_path:
+                    versions[ioc_name] = 'work'
+                else:
+                    index = ioc_path.index(ioc_name)
+                    end = ioc_path[index+len(ioc_name)+1:]
+                    version = end.split(os.path.sep)[0]
+                    versions[ioc_name] = version
```
Comment (nickbattam) [inline html_versions.py:223]: not really any need to declare the 'version' variable here; this might slow down loop execution too

Comment (willrogers) [inline html_versions.py:223]: pff in this case the variable 'version' makes it clearer what's happening

File: html_versions.py (hunk 8 above)
Comment (willrogers) [inline html_versions.py:212]: Which line?  206?

**Comment by nickbattam** [comment]
Is there any way of running this analysis on a file system e.g. /dls_sw/work/R3.14.12.3, /dls_sw/prod/R3.14.12.3 or <local shadow file system>?

File: html_versions.py (hunk 8 above)
Comment (nickbattam) [inline html_versions.py:212]: 217:  `ioc_paths = [path for _, path in (line.split() for line in cfg_ioc)]`

**Comment by willrogers** [comment]
@nickbattam I've fixed most of your suggestions.  OK to merge?

**Comment by willrogers** [comment]
and now?

**Pull request merged**